### PR TITLE
[ISSUE #3094]abstract create nacos restTemplate factory

### PR DESCRIPTION
**Pull request by Maijh97**
Please do not create a Pull Request without creating an issue first.

## What is the purpose of the change
fix: #3094 
When using nacos templates,can customize common parameters of http client, such as timeout time and connection time.

- Add HttpClientFactory interface to create `NacosRestTemplate` and `NacosAsyncRestTemplate`
- Abstract the `AbstractHttpClientFactoryWrapper` class, encapsulate the method created, provide custom http client config method
- Defines a unified http client config class
- Defines the tool class creates `NacosRestTemplate` and `NacosAsyncRestTemplate`

## Brief changelog

XX

## Verifying this change

XXXX

Follow this checklist to help us incorporate your contribution quickly and easily:

* [x] Make sure there is a Github issue filed for the change (usually before you start working on it). Trivial changes like typos do not require a Github issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue.
* [x] Format the pull request title like `[ISSUE #123] Fix UnknownException when host config not exist`. Each commit in the pull request should have a meaningful subject line and body.
* [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
* [ ] Write necessary unit-test to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add integration-test in [test module](https://github.com/alibaba/nacos/tree/master/test).
* [ ] Run `mvn -B clean package apache-rat:check findbugs:findbugs -Dmaven.test.skip=true` to make sure basic checks pass. Run `mvn clean install -DskipITs` to make sure unit-test pass. Run `mvn clean test-compile failsafe:integration-test`  to make sure integration-test pass.

### CODE REVIEW

**Comment by KomachiSion** [comment]
1. Please change your title with [ISSUE #XXX] xxx. 
2. And refer your issue in your description.
3. update your finished checklist.

**Comment by Maijh97** [comment]
本来打算替换客户端的http client使用，但是发现每个使用的http client都有自己独有配置参数。为了保证PR内容尽量少，我这边暂时先不做替换的工作，先把nacos template创建重新抽象了，在这个PR合并之后，我将会继续替换客户端http client的任务，麻烦 @chuntaojun @KomachiSion 同学帮忙review该PR。谢谢

File: common/src/main/java/com/alibaba/nacos/common/http/HttpClientBeanFactory.java [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+
+/**
+ * Create a rest template
+ * to ensure that each custom client config and rest template are in one-to-one correspondence
+ *
+ * @author mai.jh
+ * @date 2020/6/16
+ */
+@SuppressWarnings("all")
+public final class HttpClientBeanFactory {
```
Comment (KomachiSion) [inline common/src/main/java/com/alibaba/nacos/common/http/HttpClientBeanFactory.java:39]: Maybe rename to `HttpClientBeanCache` or `HttpClientBeanHolder` will be better.

File: common/src/main/java/com/alibaba/nacos/common/http/HttpClientBeanFactory.java [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+            @Override
+            public void run() {
+                shutdown();
+            }
+        });
+    }
+
+    public static NacosRestTemplate getNacosRestTemplate() {
+        HttpClientFactory httpClientFactory = new DefaultHttpClientConfig();
+        return getNacosRestTemplate(httpClientFactory);
```
Comment (KomachiSion) [inline common/src/main/java/com/alibaba/nacos/common/http/HttpClientBeanFactory.java:64]: merge these line to single line to remove tmp variable.

like `return getNacosRestTemplate(new DefaultHttpClientConfig());`

File: common/src/main/java/com/alibaba/nacos/common/http/HttpClientBeanFactory.java [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+                nacosRestTemplate = httpClientFactory.createNacosRestTemplate();
+                singletonRest.put(factoryName, nacosRestTemplate);
+            }
+        }
+        return nacosRestTemplate;
+    }
+
+    public static NacosAsyncRestTemplate getNacosAsyncRestTemplate() {
+        HttpClientFactory httpClientFactory = new DefaultHttpClientConfig();
+        return getNacosAsyncRestTemplate(httpClientFactory);
```
Comment (KomachiSion) [inline common/src/main/java/com/alibaba/nacos/common/http/HttpClientBeanFactory.java:88]: Same as above description.

File: common/src/main/java/com/alibaba/nacos/common/http/AbstractHttpClientFactoryWrapper.java [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+import org.apache.http.impl.nio.client.HttpAsyncClients;
+
+/**
+ * AbstractHttpClientFactory
+ * Let the creator only specify the http client config
+ *
+ * @author mai.jh
+ * @date 2020/6/15
+ */
+public abstract class AbstractHttpClientFactoryWrapper implements HttpClientFactory {
```
Comment (KomachiSion) [inline common/src/main/java/com/alibaba/nacos/common/http/AbstractHttpClientFactoryWrapper.java:34]: `AbstractHttpClientFactory` is enough.

File: client/src/main/java/com/alibaba/nacos/client/naming/net/NamingHttpClientManager.java [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+            return "https://";
+        }
+        return "http://";
+    }
+
+    public static NacosRestTemplate getNacosRestTemplate() {
+        return HttpClientBeanFactory.getNacosRestTemplate(HTTP_CLIENT_FACTORY);
+    }
+
+    private static class NamingHttpClientConfig extends AbstractHttpClientFactoryWrapper {
```
Comment (KomachiSion) [inline client/src/main/java/com/alibaba/nacos/client/naming/net/NamingHttpClientManager.java:51]: `NamingHttpClientFactory` will be better. `NamingHttpClientConfig` make other think it is a config pojo.

File: common/src/main/java/com/alibaba/nacos/common/http/DefaultHttpClientConfig.java [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+
+package com.alibaba.nacos.common.http;
+
+/**
+ * default http client factory
+ *
+ * @author mai.jh
+ * @date 2020/6/15
+ */
+public class DefaultHttpClientConfig extends AbstractHttpClientFactoryWrapper {
```
Comment (KomachiSion) [inline common/src/main/java/com/alibaba/nacos/common/http/DefaultHttpClientConfig.java:25]: Same problem as `NamingHttpClientConfig`. `DefaultHttpClientFactory` will be better.

File: example/src/main/java/com/alibaba/nacos/example/NamingExample.java [hunk 7]
```diff
@@ -55,4 +55,6 @@ public void onEvent(Event event) {
             }
         });
     }
+
+
```
Comment (KomachiSion) [inline example/src/main/java/com/alibaba/nacos/example/NamingExample.java:59]: Why add this two blank line? Please remove.

**Pull request merged**